### PR TITLE
Introduce auth_ident_file setting and enhance cert and peer auth methods with user name maps

### LINKED ISSUE #148: Support for username map= parameter and username maps
We have auto-scaling EC2 instances that get issued certificates with Common Names containing the EC2 instance ID. Our intent is to use pgbouncer in hostssl/cert mode, but support is needed for the username map parameter in HBA (and some kind of username mapping similar to postgresql's pg_ident.conf) to map the certificate CN to a known role. Ideally, the feature would mimic postgresql's implementation (i.e. with regexp support etc).

The documentation (https://pgbouncer.github.io/config.html#hba-file-format) says:

> username map (map=) parameter is not supported.

Keen to get this feature sooner rather than later. Can any of the existing maintainers help us out?

### LINKED ISSUE #767: Certificate based authentication issues in pgBouncer 1.17.0
we are facing an issue with enabling cert based authentication between clients and pgbouncer 1.17.0. Pgbouncer 1.17.0 and postgresql 14 are on the same host in different containers. There is TLS enabled between client and pgbouncer. The issue is that we want client to connect to pgbouncer with client key and certificate, the certificate will have a Common Name (lets say "abc") which wont be same as the user "joe" which is trying to login to pgbouncer, so here we want a mapping of the user "joe" to "abc". If we try to login as joe , Immediately we get error because pgbouncer will check the user ("joe") and Common Name ("abc") in cert and they doesn't match. What is the right way of solving this issue? otherwise how can different users login to pgBouncer?

On postgreSQL, this mapping can be done using pg_ident.conf but that support is not available in pgBouncer. What is the most secured way of achieving this authentication?

### PR DESCRIPTION
**Pull request by emelsimsek**
Postgres supports user name maps[1] for `cert`  and `peer` authentication methods. This change provides the same functionality with some limitations for pgbouncer. The limitations are listed in "Ident map file format" section of the docs.

1. https://www.postgresql.org/docs/current/auth-username-maps.html

Fixes #767
Fixes #148

### CODE REVIEW

File: doc/config.md
Comment (JelteF) [inline doc/config.md:451]: Let's call this auth_ident_file to use teh same convention as hba_file (that's not called auth_pghba_file either).
```suggestion
### auth_ident_file
```

File: test/test_auth.py
Comment (JelteF) [inline test/test_auth.py:497]: These two files are not committed, which is why this test is failing.

File: test/test_auth.py
Comment (JelteF) [inline test/test_auth.py:481]: Some additional tests that should be added:
- [ ] Rejects connection with the wrong CN in the cert (e.g. try connecting with localhost cert)
- [ ] Rejects connection when CN matches, but mapname doesn't match.
- [ ] "all" allows connection with any user

File: doc/config.md
Comment (JelteF) [inline doc/config.md:1428]:
```suggestion
* Database-username field: Supports `all` or a single postgres user name. Not supported: `+groupname`, multiple names, regular expressions.
```

File: src/client.c
Comment (JelteF) [inline src/client.c:226]: This logic is incorrect in a small but very important way. W should only try a direct match if no map is defined. If a map is defined, it's very much possible that a direct match is actually not allowed to give access.

File: src/client.c
Comment (JelteF) [inline src/client.c:237]: There's two ways all can be formatted in the ident file, which means two different things:
- `all` (literal string), this is the special keyword that matches all users
- `"all"` (quoted), this means it only matches the user called `all` (assuming this user exists)

Postgres tracks this difference by setting a flag in case the special keyword is used.

File: src/main.c
Comment (JelteF) [inline src/main.c:456]: nit: this check is unnecessary, you can always free even if it is NULL (then free is a no-op)

File: src/client.c
Comment (JelteF) [reply src/client.c:237]: For HBA parsing we use the `NAME_ALL` flag in pgbouncer too actually.

File: src/hba.c
Comment (JelteF) [inline src/hba.c:516]: Afaict this logic is to handle spaces between the `=` sign an `map`. This is unnecessary, postgres also does not support this: https://github.com/postgres/postgres/blob/master/src/backend/libpq/hba.c#L1856-L1884

Also, as you can see in the `parse_hba_auth_opt` function, Postgres actually supports a bunch of these options in hba lines. Let's at least prepare our code to do the same (i.e. using `strchr` to get the option name). That way we can give more useful error messages, e.g. option clientcert is not supported.

File: doc/config.md
Comment (JelteF) [inline doc/config.md:1432]: Multiple values are not supported by Postgres either in the ident file (so let's make this limitation implicit).

```suggestion
* Database-username field: Supports `all` or a single postgres user name. Not supported: `+groupname`, regular expressions.
```

File: src/hba.c
Comment (JelteF) [inline src/hba.c:27]: Let's move this one to hba.h too, to keep these defines together.

**Comment by onurctirtir** [comment]
SIGSEGV happens when trying to connect to a database that's present in ini file:

```c
#0  0x000055a024613e61 in finish_set_pool (client=client@entry=0x55a025a09a80, takeover=takeover@entry=false) at src/client.c:310
#1  0x000055a024615918 in handle_client_startup (client=client@entry=0x55a025a09a80, pkt=pkt@entry=0x7ffc669afb10) at src/client.c:928
#2  0x000055a024616549 in client_proto (sbuf=0x55a025a09c50, evtype=<optimized out>, data=0x7ffc669afb90) at src/client.c:1365
#3  0x000055a024628f68 in sbuf_call_proto (sbuf=sbuf@entry=0x55a025a09c50, event=event@entry=0) at src/sbuf.c:498
#4  0x000055a0246294fa in sbuf_process_pending (sbuf=sbuf@entry=0x55a025a09c50) at src/sbuf.c:802
#5  0x000055a024629e24 in sbuf_main_loop (skip_recv=<optimized out>, sbuf=0x55a025a09c50) at src/sbuf.c:1031
#6  sbuf_main_loop (sbuf=0x55a025a09c50, skip_recv=<optimized out>) at src/sbuf.c:966
#7  0x000055a024616230 in handle_auth_query_response (client=0x55a025a09a80, pkt=pkt@entry=0x7ffc669b05f0) at src/client.c:545
#8  0x000055a02462cf56 in handle_server_work (pkt=0x7ffc669b05f0, server=0x55a025a23a80) at src/server.c:458
#9  server_proto (sbuf=0x55a025a23c50, evtype=<optimized out>, data=<optimized out>) at src/server.c:702
#10 0x000055a024628f68 in sbuf_call_proto (sbuf=sbuf@entry=0x55a025a23c50, event=event@entry=0) at src/sbuf.c:498
#11 0x000055a0246294fa in sbuf_process_pending (sbuf=sbuf@entry=0x55a025a23c50) at src/sbuf.c:802
#12 0x000055a024629e24 in sbuf_main_loop (skip_recv=<optimized out>, sbuf=0x55a025a23c50) at src/sbuf.c:1031
#13 sbuf_main_loop (sbuf=0x55a025a23c50, skip_recv=<optimized out>) at src/sbuf.c:966
#14 0x00007ff08f61e13f in ?? () from /lib/x86_64-linux-gnu/libevent-2.1.so.7
#15 0x00007ff08f61e87f in event_base_loop () from /lib/x86_64-linux-gnu/libevent-2.1.so.7
#16 0x000055a02461ccba in main_loop_once () at src/main.c:794
#17 0x000055a02460ec45 in main (argc=<optimized out>, argv=<optimized out>) at src/main.c:1082
```

Assuming that `ident_map` is non-NULL seems problematic at src/client.c:310:
```c
log_warning("hba_eval returned Ident map %s %s %s", rule->identmap->map_name, rule->identmap->system_user_name, rule->identmap->postgres_user_name);
```

File: src/client.c
Comment (JelteF) [inline src/client.c:226]: We should log errors here, failures in these kind of things are hard enough to debug already.

File: src/client.c
Comment (JelteF) [inline src/client.c:225]: Let's add similar logic to use the `map` when available to `login_as_unix_peer`, because postgres allow username mapping for that one too: https://www.postgresql.org/docs/current/auth-peer.html

(and afaict to support that we only similar 5 lines)

File: test/pgident.conf
Comment (JelteF) [inline test/pgident.conf:5]: I think it would be good to test a case where there are multiple mappings for the same name, then the cert should be able to authenticate as both:
```suggestion
test		pgbouncer.acme.org	someuser
test		pgbouncer.acme.org	someuser2
```

File: test/test_auth.py
Comment (JelteF) [inline test/test_auth.py:622]: sidenote: We should change this for `bouncer.pg.create_user` once #876 is merged

File: include/hba.h
Comment (JelteF) [inline include/hba.h:54]: Let's use CamelCase for the struct name. I realize took the naming scheme from HBA, but that is an acronym (and even then I don't agree with this type of casing personally).
```suggestion
struct IdentMap {
```

File: src/hba.c
Comment (JelteF) [inline src/hba.c:575]: This one should allow TOK_STRING too.

File: test/pgident.conf
Comment (JelteF) [inline test/pgident.conf:6]: We should have some tests for quoted names too in all of the columns.

File: src/hba.c
Comment (JelteF) [reply src/hba.c:565]: This one still requires TOK_IDENT for the map_name afaict.

**Review by JelteF — CHANGES_REQUESTED** [review]

File: test/pgident.conf
Comment (emelsimsek) [reply test/pgident.conf:5]: The current design does not support this.  The first encountered map is effective. But I see that postgres supports multiple mappings with the same name:
https://www.postgresql.org/docs/current/auth-username-maps.html#:~:text=%23%20MAPNAME%20%20%20%20%20%20%20SYSTEM%2DUSERNAME,omicron%20%20%20%20%20%20%20%20%20bryanh%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20guest1

Need to make data structure changes to support this.

File: src/client.c
Comment (emelsimsek) [reply src/client.c:225]: I have done this change but could not add tests for it.  I was able to manually test.

I tried adding a test via: 

 result = subprocess.run([f"**sudo -u asystemuser psql -p {bouncer.port} -h  {bouncer.admin_host} p0 -c \"SELECT 1;\"**"], stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True)
    print(result.returncode, result.stdout, result.stderr)

returns error 

..... Permission denied\npsql: error: connection to server on socket **"/tmp/pytest-of-emel/pytest-34/test_client_hba_cert0/bouncer/.s.PGSQL.10202**" failed: Permission denied\n\tIs the server running locally and accepting connections on that socket?\n'

Comment (emelsimsek) [reply src/client.c:225]: Maybe we should make this a seperate PR with proper tests added.

Comment (JelteF) [reply src/client.c:225]: I think probably the easiest way to create a test is to manually create an ident file, that contains the current system user. So something like this (untested code):
```python
cur_user = os.current_user()
with open("myident.conf", 'w') as f:
    f.write(f"mymap {cur_user} postgres")	

bouncer.write_ini(f"auth_ident_file = pgident.conf")
```

Comment (JelteF) [reply src/client.c:225]: > Maybe we should make this a seperate PR with proper tests added.

One big reason why I'd like it in too, is because it's generally much easier to do some manual testing with it.

Comment (emelsimsek) [reply src/client.c:225]: Thanks @JelteF. I pushed a simple test now.

File: test/test_auth.py
Comment (JelteF) [inline test/test_auth.py:754]: These files should be created in the config directory of pgbouncer. Otherwise they polute the current working directory when running the tests locally.

**Review by JelteF — APPROVED** [review]
LGTM now. But as already discussed in private let's wait until we merged and released #1025 as a patch release

**Pull request merged**
